docs/design/gateway/lwM2M-gateway.md

```diff
@@ -0,0 +1,211 @@
+
+## Goal
+
+### Design Goal
+
+- Enable the device to act as an LwM2M device registered with the LwM2M server.
+- Provide a method to connect deviceShifu to the LwM2M server.
+- Allow the LwM2M server to read, write, and execute instructions on deviceShifu via the LwM2M protocol.
+
+### Non-Goal
+
+- Support for all features in the LwM2M protocol.
+- Automatic recognition and generation of LwM2M Objects and Resources according to the [OMA Standard](https://github.com/OpenMobileAlliance/lwm2m-registry).
+
+### Features
+
+- Support for [LwM2M protocol v1.0.x](https://www.openmobilealliance.org/release/LightweightM2M/V1_0-20170208-A/OMA-TS-LightweightM2M-V1_0-20170208-A.pdf).
+- LwM2M protocol over UDP.
+- Datagram Transport Layer Security (DTLS) support.
+- Support for LwM2M protocol communication with the server.
+- Support for `read`, `write` and `Execute` requests.
+- Support Notify and Observe feature.
+
+### Unsupported Features
+
+-	Support for LwM2M v1.1.x or later.
+-	Support over TCP or other protocols.
+-	Bootstrap server.
+-	Support for all LwM2M Objects.
+-	Support for all LwM2M Resources.
+
+## LwM2M Gateway Design
+
+The LwM2M Gateway consists of two main components: an LwM2M client that connects to the LwM2M server and an HTTP client that connects to deviceShifu.
+
+When the server sends a read or write request to the gateway, the gateway calls the deviceShifu instructions to either retrieve or set the data. If the server enables the Observe feature, the gateway regularly collects data from the deviceShifu. If the data changes or a timeout occurs, the gateway notifies the server with the updated data.
+
+```mermaid
+flowchart BT
+
+ls[LwM2M-server]
+
+subgraph EdgeNode
+    subgraph Shifu
+      subgraph ds1[deviceShifu-HTTP]
+          dsh[deviceShifu-HTTP]
+          gl1[LwM2M-gateway]
+          dsh <-->|HTTP| gl1
+      end
+      subgraph ds2[deviceShifu-MQTT]
+          dsm[deviceShifu-MQTT]
+          gl2[LwM2M-gateway]
+          dsm <-->|HTTP| gl2
+      end
+        subgraph ds3[deviceShifu-LwM2M]
+          dsl[deviceShifu-LwM2M]
+          gl3[LwM2M-gateway]
+          dsl <-->|HTTP| gl3
+      end
+    end
+end
+
+
+dh[device-HTTP]
+dm[device-MQTT]
+dl[device-LwM2M]
+
+
+dh -->|HTTP| dsh
+gl1 -->|LwM2M| ls
+
+dm -->|MQTT| dsm
+gl2 -->|LwM2M| ls
+
+dl -->|LwM2M| dsl
+gl3 -->|LwM2M| ls 
+```
+
+### What Does the Gateway Do?
+
+1.	Start the LwM2M client and obtain all device information from deviceShifu.
+2.	Register the device with the server and update the device information.
+3.	Handle requests from the server.
+4.	If the server enables the Observe feature, notify the server when data changes or a timeout occurs.
```

Comment:
which timeout? do you mean there is an interval can be configured?

Reply:
The interval is configured by the LwM2M server, but it resets whenever the data changes. i think interval is not good for this